dot/rpc/websocket.go

```diff
@@ -38,59 +91,147 @@ func (h *HTTPServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		log.Error("[rpc] websocket upgrade failed", "error", err)
 		return
 	}
+
+	rpcHost := fmt.Sprintf("http://%s:%d/", h.serverConfig.Host, h.serverConfig.RPCPort)
 	for {
-		rpcHost := fmt.Sprintf("http://%s:%d/", h.serverConfig.Host, h.serverConfig.RPCPort)
-		for {
-			_, mbytes, err := ws.ReadMessage()
-			if err != nil {
-				log.Error("[rpc] websocket failed to read message", "error", err)
-				return
+		_, mbytes, err := ws.ReadMessage()
+		if err != nil {
+			log.Error("[rpc] websocket failed to read message", "error", err)
+			return
+		}
+		log.Trace("[rpc] websocket received", "message", fmt.Sprintf("%s", mbytes))
+
+		// determine if request is for subscribe method type
+		var msg map[string]interface{}
+		err = json.Unmarshal(mbytes, &msg)
+		if err != nil {
+			log.Error("[rpc] websocket failed to unmarshal request message", "error", err)
+			res := &ErrorResponseJSON{
+				Jsonrpc: "2.0",
+				Error: &ErrorMessageJSON{
+					Code:    big.NewInt(-32600),
+					Message: "Invalid request",
+				},
+				ID: nil,
 			}
-			log.Trace("[rpc] websocket received", "message", fmt.Sprintf("%s", mbytes))
-			client := &http.Client{}
-			buf := &bytes.Buffer{}
-			_, err = buf.Write(mbytes)
+			err = ws.WriteJSON(res)
 			if err != nil {
-				log.Error("[rpc] failed to write message to buffer", "error", err)
-				return
+				log.Error("[rpc] websocket failed write message", "error", err)
+			}
+			continue
+		}
+		method := msg["method"]
+		// if method contains subscribe, then register subscription
+		if strings.Contains(fmt.Sprintf("%s", method), "subscribe") {
+			mid := msg["id"].(float64)
```

Comment:
is the subscription ID provided by the user?

Reply:
No, the id provide by the user is the message id, so that can track that to the responses.  The subscription id is created by the service and returned to the requester.

Reply:
ah okay makes sense!